Firestore/core/src/firebase/firestore/remote/grpc_completion.h

```diff
@@ -120,6 +138,37 @@ class GrpcCompletion {
 
   void EnsureValidFuture();
 
+  // GrpcCompletions are meant to be created and passed to gRPC as a context
+  // object. gRPC's API only allows for a raw pointer though, and until gRPC
+  // invokes the callback with this GrpcCompletion as a context object the
+  // object must not be deleted.
+  //
+  // Previously we managed this by explicitly new/deleting GrpcCompletion but
+  // this creates timing problems during shutdown. During shutdown, the worker
+  // thread wants to cancel RPCs that are in flight, but this races with the
+  // completion event from gRPC.
+  //
+  // This used to work because we only allowed delete on the worker thread.
+  // Now that the AsyncQueue no longer admits tasks after shutdown has been
+  // started this now leaks the completion. If we called delete outside the
```

Comment:
I initially considered this as a possibility but think it's problematic. EnqueueEvenAfterShutdown is a hack I'd really like to get rid of and going doing that path would likely cement it into place. I can be convinced otherwise, but this solution is viable, and changing this again/reviewing again will make us miss the release window, which I don't want to do.